src/lib/vasi-sync/src/scchannel.rs

```diff
@@ -160,6 +161,7 @@ impl Display for SelfContainedChannelError {
 /// microbenchmark; the latter is expected to be more representative of real
 /// large simulation runs (i.e. pinning should be enabled).
 #[cfg_attr(not(loom), derive(VirtualAddressSpaceIndependent))]
+#[repr(C)]
 pub struct SelfContainedChannel<T> {
     message: UnsafeCell<MaybeUninit<T>>,
     state: AtomicChannelState,
```

Comment:
Shouldn't this have been enforced by the `_validate_stable_layout` in the `ShMemBlock`?

Reply:
I think that might only catch if the top-level type has a non-stable repr.

Reply:
which in this case is `IPCData`

Reply:
Actually I wonder if we could get a recursive check by moving `_validate_stable_layout` into one of the marker traits. will experiment with that...

Reply:
Adding the function to the trait doesn't work; trait methods can't take `self`, only `&self`, which doesn't forbid `Self` from being `[repr(rust)]`.

Maybe the derive macro could *inject* the `_validate_stable_layout` method?

Reply:
I think it is recursive, but it's not working because it's generic.  For example the following successfully fails to build:

```rust
#[repr(C)]
struct Test(TestInner);

#[repr(C)]
struct TestInner(Vec<i32>);

#[deny(improper_ctypes_definitions)]
extern "C" fn _test(_x: Test) {}
```

But in `ShMemBlock` it uses a generic. For example the following does build without errors:

```
#[repr(C)]
struct Test(TestInner);

#[repr(C)]
struct TestInner(Vec<i32>);

#[deny(improper_ctypes_definitions)]
extern "C" fn test<T>(_x: T) {}

fn main() {
    test(Test(TestInner(Vec::new())));
}
```

Reply:
> Maybe the derive macro could inject the `_validate_stable_layout` method?

This is the only way I can think of to make this work. I think we could add this to the `VirtualAddressSpaceIndependent` derive and specify the type in the signature explicitly, or maybe make a new derive macro named something like `AssertStableLayout` which does this, and then `ShmMemBlock` can require that `T: VirtualAddressSpaceIndependent + AssertStableLayout`.

Anyway this isn't needed for this PR, but we should probably fix this so that it does detect non-stable layouts in shared memory, or remove the existing `_validate_stable_layout` that doesn't seem to work.

Reply:
I tried doing this locally but it didn't seem to catch the error. Oh well, I'll open another PR for follow-up